pkg/localstorev2/internal/pinning/pinning.go

```diff
@@ -0,0 +1,281 @@
+	}
+	if len(p.UUID) == 0 {
+		return nil, errInvalidPinCollectionUUID
+	}
+	buf := p.Addr.Bytes()
+	buf = append(buf, []byte(p.UUID)...)
+	statBuf := make([]byte, 8+8)
+	binary.LittleEndian.PutUint64(statBuf[:8], p.Stat.Total)
+	binary.LittleEndian.PutUint64(statBuf[8:16], p.Stat.DupInCollection)
+	buf = append(buf, statBuf...)
+	return buf, nil
+}
+
+func (p *pinCollectionItem) Unmarshal(buf []byte) error {
+	if len(buf) != pinCollectionItemSize {
+		return errInvalidPinCollectionSize
+	}
+	ni := new(pinCollectionItem)
+	ni.Addr = swarm.NewAddress(buf[:swarm.HashSize])
+	ni.UUID = string(buf[swarm.HashSize : swarm.HashSize+uuidSize])
+	statBuf := buf[swarm.HashSize+uuidSize:]
+	ni.Stat.Total = binary.LittleEndian.Uint64(statBuf[:8])
+	ni.Stat.DupInCollection = binary.LittleEndian.Uint64(statBuf[8:16])
+	*p = *ni
+	return nil
+}
+
+// pinChunkItem is the index used to represent a single chunk in the pinning
+// collection. It is prefixed with the UUID of the collection.
+type pinChunkItem struct {
+	UUID string
+	Addr swarm.Address
+}
+
+func (p *pinChunkItem) Namespace() string { return p.UUID }
+
+func (p *pinChunkItem) ID() string { return p.Addr.ByteString() }
+
+// pinChunkItem is a key-only type index. We don't need to store any value. As such
+// the serialization functions would be no-ops. A Get operation on this key is not
+// required as the key would constitute the item. Usually these type of indexes are
+// useful for key-only iterations.
+func (p *pinChunkItem) Marshal() ([]byte, error) {
+	return []byte{}, nil
+}
+
+func (p *pinChunkItem) Unmarshal(_ []byte) error {
+	return nil
+}
+
+// PutterCloser implements the storage.Putter as well as an io.Closer. The Putter
+// should be closed for a successful operation.
+type PutterCloser interface {
+	storage.Putter
+	io.Closer
+}
+
+// Storage interface is the functionality required from the underlying storage to
+// achieve the pinstore functionality. This allows us to do stateless operations
+// here.
+type Storage interface {
+	Ctx() context.Context
+	Store() storage.Store
+	ChunkStore() storage.ChunkStore
+}
+
+// NewCollection returns a putter wrapped around the passed storage.
+// The putter will add the chunk to Chunk store if it doesnt exists within this collection.
+// It will create a new UUID for the collection which can be used to iterate on all the chunks
+// that are part of this collection. The root pin is only updated on successful close of this
+// Putter.
+func NewCollection(st Storage, root swarm.Address) (PutterCloser, error) {
+
+	collection := &pinCollectionItem{Addr: root, UUID: uuid.NewString()}
+	found, err := st.Store().Has(collection)
+	if err != nil {
+		return nil, fmt.Errorf("pin store: failed checking collection: %w", err)
+	}
+	if found {
+		return nil, fmt.Errorf("pin store: root %s already exists", root)
+	}
+
+	return &collectionPutter{collection: collection, st: st}, nil
+}
+
+type collectionPutter struct {
+	mtx        sync.Mutex
```

Comment:
Not sure why this mutex is needed, is it only because of the `Close`?

Reply:
Yes